src/client.cpp

```diff
@@ -1386,3 +1458,179 @@ int CClient::EstimatedOverallDelay ( const int iPingTimeMs )
+        // all other fields will be initialised on channel allocation
+
+        clientChannelIDs[i] = INVALID_INDEX;
+    }
+
+    qInfo() << "> Client channel list cleared";
+}
+
+void CClient::FreeClientChannel ( const int iServerChannelID )
+{
+    QMutexLocker locker ( &MutexChannels );
+
+    if ( iServerChannelID == INVALID_INDEX || iServerChannelID >= MAX_NUM_CHANNELS )
+    {
+        return;
+    }
+
+    const int iClientChannelID = clientChannelIDs[iServerChannelID];
+
+    Q_ASSERT ( clientChannels[iClientChannelID].iServerChannelID == iServerChannelID );
+
+    clientChannelIDs[iServerChannelID]                = INVALID_INDEX;
+    clientChannels[iClientChannelID].iServerChannelID = INVALID_INDEX;
+
+    iActiveChannels -= 1;
+
+    qInfo() << qUtf8Printable ( QString ( "> Freed client ch %1 for server ch %2; chan count = %3" )
+                                    .arg ( iClientChannelID )
+                                    .arg ( iServerChannelID )
+                                    .arg ( iActiveChannels ) );
+}
+
+// find, and optionally create, a client channel for the supplied server channel ID
+// returns a client channel ID or INVALID_INDEX
+int CClient::FindClientChannel ( const int iServerChannelID, const bool bCreateIfNew )
+{
+    QMutexLocker locker ( &MutexChannels );
+
+    if ( iServerChannelID == INVALID_INDEX || iServerChannelID >= MAX_NUM_CHANNELS )
+    {
+        return INVALID_INDEX;
+    }
+
+    int iClientChannelID = clientChannelIDs[iServerChannelID];
+
+    if ( iClientChannelID != INVALID_INDEX )
+    {
+        Q_ASSERT ( clientChannels[iClientChannelID].iServerChannelID == iServerChannelID );
+
+        return iClientChannelID;
+    }
+
+    // no matching client channel - create new one if requested
+    if ( bCreateIfNew )
+    {
+        // search clientChannels[] for a free client channel
+        for ( iClientChannelID = 0; iClientChannelID < MAX_NUM_CHANNELS; iClientChannelID++ )
+        {
+            CClientChannel* clientChan = &clientChannels[iClientChannelID];
+
+            if ( clientChan->iServerChannelID == INVALID_INDEX )
+            {
+                clientChan->iServerChannelID = iServerChannelID;
+                clientChan->iJoinSequence    = ++iJoinSequence;
+
+                clientChan->oldGain = clientChan->newGain = 1.0f;
+
+                clientChan->level = 0;
+
+                clientChannelIDs[iServerChannelID] = iClientChannelID;
+
+                iActiveChannels += 1;
+
+                qInfo() << qUtf8Printable ( QString ( "> Alloc client ch %1 for server ch %2; chan count = %3" )
+                                                .arg ( iClientChannelID )
+                                                .arg ( iServerChannelID )
+                                                .arg ( iActiveChannels ) );
+
+                return iClientChannelID; // new client channel ID
+            }
+        }
+    }
+
+    return INVALID_INDEX;
```

Comment:
Are there any _normal_ reasons to get here?  I'd log again...

Either we expect the channel to exist, so do not request one created -- so INVALID_INDEX is very unexpected.

Or we did expected the channel might not exist, so we requested a new one -- but we got more server channels than the client knows about, so INVALID_INDEX is highly unlikely (as we generally have newer clients than servers, I'd think).

Reply:
The second parameter to the function specifies whether to create a new client channel or not. There are a few places where `FindClientChannel` is called with `false`, in which case we want the `INVALID_INDEX` return for an unallocated channel.